text/0000-variance.md

```diff
@@ -0,0 +1,568 @@
+parameters would always be used, but this is not true. In particular,
+some types have "phantom" type or lifetime parameters that are not
+used in the body of the type. This generally occurs with unsafe code:
+
+    struct Items<'vec, T> { // unused lifetime parameter 'vec
+        x: *mut T
+    }
+    
+    struct AtomicPtr<T> { // unused type parameter T
+        data: AtomicUint  // represents an atomically mutable *mut T, really
+    }
+    
+Since these parameters are unused, the inference can reasonably
+conclude that `AtomicPtr<int>` and `AtomicPtr<uint>` are
+interchangable: after all, there are no fields of type `T`, so what
+difference does it make what value it has? This is not good (and in
+fact we have behavior like this today for lifetimes, which is a common
+source of error).
+
+To avoid this hazard, the RFC proposes to make it an error to have a
+type or lifetime parameter whose variance is not constrained. Almost
+always, the correct thing to do in such a case is to either remove the
+parameter in question or insert a *marker type*. Marker types
+basically inform the inference engine to pretend as if the type
+parameter were used in particular ways. They are discussed in the next section.
+
+## Revamping the marker types
+
+### The UnsafeCell type
+
+As today, the `UnsafeCell<T>` type is well-known to `rustc` and is
+always considered invariant with respect to its type parameter `T`.
+
+### Phantom data
+
+This RFC proposes to replace the existing marker types
+(`CovariantType`, `ContravariantLifetime`, etc) with a single type,
+`PhantomData`:
+
+```rust
+// Represents data of type `T` that is logically present, although the
+// type system cannot see it. This type is covariant with respect to `T`.
+struct PhantomData<T>;
+```
+    
+An instance of `PhantomData` is used to represent data that is
+logically present, although the type system cannot see
+it. `PhantomData` is covariant with respect to its type parameter `T`. Here are
+some examples of uses of `PhantomData` from the standard library:
+    
+```rust    
+struct AtomicPtr<T> {
+    data: AtomicUint,
+    
+    // Act as if we could reach a `*mut T` for variance. This will
+    // make `AtomicPtr` *invariant* with respect to `T` (because `T` appears
+    // underneath the `mut` qualifier).
+    marker: PhantomData<*mut T>,
+}
+    
+pub struct Items<'a, T: 'a> {
+    ptr: *const T,
+    end: *const T,
+    
+    // Act as if we could reach a slice `[T]` with lifetime `'a`.
+    // Induces covariance on `T` and suitable variance on `'a`
+    // (covariance using the definition from rfcs#391).
+    marker: marker::PhantomData<&'a [T]>,
+}
+```
+
+Note that `PhantomData` can be used to induce covariance, invariance, or contravariance
+as desired:
+
+```rust
+PhantomData<T>         // covariance
+PhantomData<*mut T>    // invariance, but see "unresolved question"
+PhantomData<Cell<T>>   // invariance
+PhantomData<fn() -> T> // contravariant
```

Comment:
Shouldn't this be covariant? `fn() -> T` is roughly isomorphic to `T`, so the result should be the same as if `PhantomData<T>` were used. Alternatively, this could be changed to `fn(T)`.